[cli] Add `vercel build` command

TooTallNate opened the pull request:
This PR adds the latest iteration of the `vc build` command which invokes the Builders that a Project depends on locally and serializes the output into the `.vercel/output` directory.

TooTallNate committed:
[cli] WIP `vc build2`

TooTallNate committed:
[build-utils] Add Builder TypeScript types

Adds formal type interfaces for Builders into `@vercel/build-utils`, and
updates the serverless functions runtimes to utilize them. This provides
type safety for the args/return values in i.e. the `build()` functions,
where previously they were not using any explicit return type.

Also removes `watch` since this is no longer being used.

TooTallNate committed:
Merge branch 'add/build-utils-builder-types' into add/cli-build2

TooTallNate committed:
.

TooTallNate committed:
.

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
.

TooTallNate committed:
.

TooTallNate committed:
.

TooTallNate committed:
Rename `vc build2` to `vc build`

TooTallNate committed:
Refactor build output to conform to new FS API spec

TooTallNate committed:
Add `zipBuffer` unzipping

TooTallNate committed:
.

TooTallNate committed:
Update `--prebuilt` to upload ".vercel/output"

TooTallNate committed:
Write Lambda config alongside and start Prerender output

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Add `Prerender` support

TooTallNate committed:
Fix relative `fsPath` for Prerender fallback

TooTallNate committed:
.

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Share meta object

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Fix build

TooTallNate committed:
Type fixes

TooTallNate committed:
Refactor Builder importing logic to use `.vercel/builders` and CLI supplied Builders

TooTallNate committed:
.

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Add redwood and static-build deps

TooTallNate committed:
[client] Don't consider `.vercelignore` for --prebuilt

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Fix redwood / static-build versions

TooTallNate committed:
Add `meta.cliVersion`

TooTallNate committed:
Alpha

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Expand `vercel.json` builds

TooTallNate committed:
Handle semver "version" and "range" when importing Builders

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Move redwood and static-build to regular deps

TooTallNate committed:
Fix redwood/static-build versions to match monorepo

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into add/cli-build2

TooTallNate committed:
Remove unused `builders-to-add.ts`

TooTallNate committed:
Load `now.json`

TooTallNate committed:
Properly consider `projectSettings.rootDirectory`

TooTallNate committed:
Don't consider `rootDirectory` for --prebuilt buildFileTree()

TooTallNate committed:
Support fully static deployment

TooTallNate committed:
Merge final `config.json` at the end and support `vercel.json` routing

TooTallNate committed:
Move `routes` back to top-level of `config.json`

TooTallNate committed:
Add Images and Wildcard types and fix wildcard

TooTallNate committed:
Explicitly handle `cleanUrls: true`

TooTallNate committed:
Trailing newline on `vc build` output

TooTallNate committed:
Add `nodeVersion` and `installCommand` to ".vercel/project.json"

TooTallNate committed:
Fix a couple TODOs and use `mime-types` for file extension lookup

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Add Builder require `name` and `apiVersion` to `builds.jsons

This will be used by `prepareCache()` to invoke the Builders'
`prepareCache()` functions.

TooTallNate committed:
Add builder import debug logging and fix post-install bug edge case

Something related to Node.js module resolution caching was making the
Builder not be resolved after it has been installed (only when
`.vercel/builders` already previously existed), so require the explicit
"main" entrypoint filename of the Builder as a workaround.

TooTallNate committed:
Serialize `EdgeFunction` output type

TooTallNate committed:
Get `nodeVersion` and `installCommand`

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Load env vars from `.vercel/.env.${target}.local` or `.env`

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Move `target` property to `builds.json`

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Fix build

TooTallNate committed:
Merge `config.json` with produced final version

TooTallNate committed:
Use `output.debug()`

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
More proper `rootDirectory` handling

TooTallNate committed:
Don't output empty `overrides` object

TooTallNate committed:
Remove a couple old TODO comments

TooTallNate committed:
Don't output HTML content-type for cleanUrls override

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Remove Lambda "type" and add EdgeFunction "runtime: 'edge'"

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

EndangeredMassa committed:
[cli] check isTTY before prompting to `pull` (#7696)

In a CI or other non-TTY setting, the `build` command should
not prompt for executing `pull`. This causes the process to
hang forever.

phil-lgr commented:
@TooTallNate this is a huge feature for a lot of nextjs users, let me know if you need help to try this out 

So far I've checkout your branch and aliased it locally, somehow the command replaces my next.config.js file with this:

![image](https://user-images.githubusercontent.com/9644867/164535498-ac1e23d0-b9cf-45f9-8342-ebf28503b761.png)

And it creates this other file:

![image](https://user-images.githubusercontent.com/9644867/164535573-70867da6-22d6-4880-857f-5834789c4a3f.png)

TooTallNate commented:
@phil-lgr Stay tuned. We're getting Next.js sorted out in the next week or so.

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Sync redwood / static-build

TooTallNate committed:
Use `fs.copy()`

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Update `yarn.lock`

TooTallNate committed:
Refactor import Builders logic to work when Builder's `package.json` defines "exports"

To fix this issue:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './package.json' is not defined
by "exports" in /project/.vercel/builders/node_modules/@nuxtjs/vercel-builder/package.json
```

TooTallNate committed:
Move comments

TooTallNate committed:
Fix `ts-eager`

TooTallNate committed:
Add temporary band-aid fix for `vercel-php`

Include this band-aid for as long as it takes for users
to update to the fixed version of `vercel-php`.

Proper fix is here: https://github.com/juicyfx/vercel-php/pull/232

TooTallNate committed:
Add "license" and "private" fields to Builders `package.json` file

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Don't override `outputDirectory` from detectBuilders()

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Support legacy `@now` scoped Builders

TooTallNate committed:
Revert "Support legacy `@now` scoped Builders"

This reverts commit af76ea78389b6d9ad96bce51c5ee86c07b5eb585.

TooTallNate committed:
Reject Builders that depend on "@now/build-utils"

TooTallNate committed:
Ignore `vc build` when "scope" is supplied

TooTallNate committed:
Add `NOW_BUILDER` legacy env var

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

TooTallNate committed:
Merge branch 'main' into add/cli-build2

TooTallNate committed:
Merge branch 'main' into add/cli-build2

TooTallNate committed:
Merge branch 'main' of https://github.com/vercel/vercel into add/cli-build2

EndangeredMassa commented on packages/cli/src/util/build/import-builders.ts:
```diff
@@ -0,0 +1,197 @@
+        }
+      }
+    }
+  } while (results.size !== builderSpecs.size);
```
I *think* I follow this looping logic, but it's not easy. I think this should be made clearer at some point. I'd be happy to pair on it.

styfle replied:
I think the confusing part is that the `do {}` is so big so its not obvious its a do-while loop

EndangeredMassa replied:
Refactored here: https://github.com/vercel/vercel/pull/7842

EndangeredMassa commented on packages/cli/src/util/build/write-build-result.ts:
```diff
@@ -0,0 +1,373 @@
+  const dest = join(OUTPUT_DIR, 'static', fsPath);
+  await fs.mkdirp(dirname(dest));
+
+  // TODO: handle (or skip) symlinks?
```
Should we make a card for this?

EndangeredMassa replied:
https://linear.app/vercel/issue/BUI-122/handle-symlinks-when-writing-build-results

EndangeredMassa reviewed (approved):
We'll want to add some tests, too.

TooTallNate commented on packages/cli/src/types.ts:
```diff
@@ -251,9 +251,6 @@ export interface Project extends ProjectSettings {
   alias?: ProjectAliasTarget[];
-  devCommand?: string | null;
-  framework?: string | null;
-  rootDirectory?: string | null;
```
They're defined in `ProjectSettings` which this interface is extending.

TooTallNate committed:
Backtick

TooTallNate commented on packages/cli/src/commands/build.ts:
```diff
@@ -0,0 +1,536 @@
+  // Merge existing `config.json` file into the one that will be produced
+  const configPath = join(OUTPUT_DIR, 'config.json');
+  // TODO: properly type
+  const existingConfig = await readJSONFile<any>(configPath);
```
Card, and also AJV validation would be good to have as well.

EndangeredMassa replied:
https://linear.app/vercel/issue/BUI-120/type-and-validate-config-to-configjson

TooTallNate commented on packages/cli/src/util/build/import-builders.ts:
```diff
@@ -0,0 +1,197 @@
+          continue;
+        }
+
+        // TODO: handle `parsed.type === 'tag'` ("latest" vs. anything else?)
```
Card is fine I think.

EndangeredMassa replied:
https://linear.app/vercel/issue/BUI-121/support-resolving-builders-by-tag

TooTallNate commented on packages/cli/src/commands/build.ts:
```diff
@@ -0,0 +1,536 @@
+
+  // Write out the final `config.json` file based on the
+  // user configuration and Builder build results
+  // TODO: properly type
```
We should define a `BuildOutputConfig` type or such, probably live in build-utils so that we can share the type with the server-side as well.

EndangeredMassa replied:
https://linear.app/vercel/issue/BUI-120/type-and-validate-config-to-configjson

TooTallNate committed:
Uncomment a couple errors

TooTallNate committed:
Remove commented code

TooTallNate commented on packages/cli/src/commands/build.ts:
```diff
@@ -0,0 +1,536 @@
+  //    link: 'https://vercel.com/docs/configuration#project/builds',
+  //    action: 'View Documentation',
+  //  });
+  //}
```
Will remove for now. We can add back later if needed.

TooTallNate committed:
Remove commented code

phil-lgr commented:
Sorry for the noise, but, could you link relevant GH issues or Vercel documentation for us to follow? Thanks.

styfle commented:
https://vercel.com/docs/cli#commands/build

```
vercel build --help
```
